pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/creator/impl/DefaultIndexCreatorProvider.java

```diff
@@ -0,0 +1,266 @@
+  public JsonIndexCreator newJsonIndexCreator(IndexCreationContext context)
+      throws IOException {
+    Preconditions.checkState(context.getFieldSpec().isSingleValueField(),
+        "Json index is currently only supported on single-value columns");
+    Preconditions.checkState(context.getFieldSpec().getDataType().getStoredType() == FieldSpec.DataType.STRING,
+        "Json index is currently only supported on STRING columns");
+    return context.isOnHeap() ? new OnHeapJsonIndexCreator(context.getIndexDir(), context.getFieldSpec().getName())
+        : new OffHeapJsonIndexCreator(context.getIndexDir(), context.getFieldSpec().getName());
+  }
+
+  @Override
+  public TextIndexCreator newTextIndexCreator(IndexCreationContext context)
+      throws IOException {
+    // Initialize text index creator
+    Preconditions.checkState(context.getFieldSpec().getDataType().getStoredType() == FieldSpec.DataType.STRING,
+        "Text index is currently only supported on STRING type columns");
+    return new LuceneTextIndexCreator(context.getFieldSpec().getName(), context.getIndexDir(),
+        true /* commitOnClose */);
+  }
+
+  @Override
+  public TextIndexCreator newFSTIndexCreator(IndexCreationContext context)
+      throws IOException {
+    Preconditions.checkState(context.getFieldSpec().isSingleValueField(),
+        "FST index is currently only supported on single-value columns");
+    Preconditions.checkState(context.getFieldSpec().getDataType().getStoredType() == FieldSpec.DataType.STRING,
+        "FST index is currently only supported on STRING type columns");
+    Preconditions.checkState(!context.isRawColumn(),
+        "FST index is currently only supported on dictionary-encoded columns");
+    String[] sortedValues = (String[]) context.getSortedUniqueElementsArray();
+    if (context.getFstType() == FSTType.NATIVE) {
+      return new NativeFSTIndexCreator(context.getIndexDir(), context.getFieldSpec().getName(), sortedValues);
+    } else {
+      return new LuceneFSTIndexCreator(context.getIndexDir(), context.getFieldSpec().getName(), sortedValues);
+    }
+  }
+
+  @Override
+  public GeoSpatialIndexCreator newGeoSpatialIndexCreator(IndexCreationContext context)
+      throws IOException {
+    Preconditions.checkState(context.getFieldSpec().isSingleValueField(),
+        "H3 index is currently only supported on single-value columns");
+    Preconditions.checkState(context.getFieldSpec().getDataType().getStoredType() == FieldSpec.DataType.BYTES,
+        "H3 index is currently only supported on BYTES columns");
+    H3IndexResolution resolution = Objects.requireNonNull(context.getH3IndexConfig()).getResolution();
+    return context.isOnHeap() ? new OnHeapH3IndexCreator(context.getIndexDir(), context.getFieldSpec().getName(),
+        resolution) : new OffHeapH3IndexCreator(context.getIndexDir(), context.getFieldSpec().getName(), resolution);
+  }
+
+  public static boolean shouldDeriveNumDocsPerChunk(String columnName,
```

Comment:
Are we planning to refactor the code in SegmentColumnarIndexCreator to avoid duplicate code and let it reuse these static methods for `shouldDeriveNumDocsPerChunk` and `rawIndexWriterVersion` ? Similarly `getRawIndexCreatorForSVColumn` can also be written just once ?

Reply:
Ignore. Saw the changes in `SegmentColumnarIndexCreator` after this and this is no longer a question